drivers/flash/soc_flash_numaker.c

```diff
@@ -0,0 +1,290 @@
+	uint32_t addr = dev_data->flash_block_base + offset;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* Validate range */
+	if (!flash_numaker_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* check alignment and erase only by pages */
+	if (((addr % FMC_FLASH_PAGE_SIZE) != 0) || ((len % FMC_FLASH_PAGE_SIZE) != 0)) {
+		return -EINVAL;
+	}
+
+	/* take semaphore */
+	if (k_sem_take(&dev_data->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
+
+	SYS_UnlockReg();
+	key = irq_lock();
+	while (page_nums) {
+		if (((len >= FMC_BANK_SIZE)) && ((addr % FMC_BANK_SIZE) == 0)) {
+			if (FMC_EraseBank(addr)) {
+				LOG_ERR("Erase flash bank failed or erase time-out");
+				rc = -EIO;
+				goto move_exit;
+			}
+			page_nums -= (FMC_BANK_SIZE / FMC_FLASH_PAGE_SIZE);
+			addr += FMC_BANK_SIZE;
+		} else {
+			/* erase page */
+			if (FMC_Erase(addr)) {
+				LOG_ERR("Erase flash page failed or erase time-out");
+				rc = -EIO;
+				goto move_exit;
+			}
+			page_nums--;
+			addr += FMC_FLASH_PAGE_SIZE;
+		}
+	}
+
+move_exit:
+	SYS_LockReg();
+	irq_unlock(key);
+	/* release semaphore */
+	k_sem_give(&dev_data->write_lock);
+
+	return rc;
+}
+
+/*
+ * Read a flash memory area.
+ *
+ * param dev       Device struct
+ * param offset    The address's offset
+ * param data      The buffer to store or read the value
+ * param length    The size of the buffer
+ * return 0       on success,
+ * return -EIO     erroneous code
+ */
+static int flash_numaker_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t addr = dev_data->flash_block_base + offset;
+
+	/* return SUCCESS for len == 0 (required by tests/drivers/flash) */
+	if (!len) {
+		return 0;
+	}
+
+	/* Validate range */
+	if (!flash_numaker_is_range_valid(offset, len)) {
+		return -EINVAL;
+	}
+
+	/* read flash */
+	memcpy(data, (void *)addr, len);
+
+	return 0;
+}
+
+static int32_t flash_numaker_block_write(uint32_t u32_addr, uint8_t *pu8_data, int block_size)
+{
+	int32_t retval;
+	uint32_t *pu32_data = (uint32_t *)pu8_data;
+
+	SYS_UnlockReg();
+	if (block_size == 4) {
+		retval = FMC_Write(u32_addr, *pu32_data);
+	} else if (block_size == 8) {
+		retval = FMC_Write8Bytes(u32_addr, *pu32_data, *(pu32_data + 1));
+	} else {
+		retval = -1;
+	}
+	SYS_LockReg();
+
+	return retval;
+}
+
+static int flash_numaker_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	struct flash_numaker_data *dev_data = dev->data;
+	uint32_t rc = 0;
+	unsigned int key;
+	uint32_t addr = dev_data->flash_block_base + offset;
+	int block_size = flash_numaker_parameters.write_block_size;
+	int blocks = (len / flash_numaker_parameters.write_block_size);
+	uint8_t *pu8_data = (uint8_t *)data;
+#if DT_NODE_HAS_PROP(SOC_NV_FLASH_NODE, write_block_size)
+	uint8_t buffer[DT_PROP(SOC_NV_FLASH_NODE, write_block_size)];
+#else
+	uint8_t buffer[0x04];
+#endif
```

Comment:
I see two points where you use this, maybe to do a #define with a default value and use it in both?

Reply:
Thanks of your advice.
I define the value and use it in both.
Above modifications are in commit: 30d523233e9